src/index/addrindex.cpp

```diff
@@ -0,0 +1,225 @@
+    bool ReadAddrIndex(const uint64_t addr_id,
+                       std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_found,
+                       const bool filter_by_value = false,
+                       const uint64_t value_wanted = 0);
+
+    bool WriteToIndex(const std::vector<std::pair<uint64_t, CDiskTxPos>> &positions, const uint256 block_hash);
+
+    void RemoveKeys(const std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_to_remove);
+};
+
+AddrIndex::DB::DB(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(GetDataDir() / "indexes" / "addrindex", n_cache_size, f_memory, f_wipe)
+{}
+
+BaseIndex::DB& AddrIndex::GetDB() const { return *m_db; }
+
+bool AddrIndex::DB::ReadAddrIndex(const uint64_t addr_id,
+                                  std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_found,
+                                  const bool filter_by_value,
+                                  const uint64_t value_wanted){
+    bool found_tx = false; // return true only if at least one transaction was found
+    const std::pair<char, uint64_t> key_prefix = std::make_pair(DB_ADDRINDEX, addr_id);
+    std::unique_ptr<CDBIterator> iter(NewIterator());
+
+    iter->Seek(key_prefix);
+    while (iter->Valid()) {
+        std::pair<std::pair<char, uint64_t>, CDiskTxPos> key;
+        uint64_t value;
+        if (!iter->GetKey(key) || !iter->GetValue(value) || key.first != key_prefix) break;
+
+        if  (!filter_by_value || (filter_by_value && value == value_wanted)) {
+            found_tx = true;
+            keys_found.emplace_back(key);
+        }
+
+        iter->Next();
+    }
+
+    return found_tx;
+}
+
+void AddrIndex::DB::RemoveKeys(const std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_to_remove) {
+    CDBBatch batch(*this);
+    for (const auto& key : keys_to_remove) {
+        batch.Erase(key);
+    }
+    WriteBatch(batch);
+}
+
+AddrIndex::AddrIndex(size_t n_cache_size, bool f_memory, bool f_wipe)
+    : m_db(MakeUnique<AddrIndex::DB>(n_cache_size, f_memory, f_wipe)){}
+
+AddrIndex::~AddrIndex() {}
+
+uint64_t AddrIndex::GetAddrID(const CScript& script) {
+    uint256 hashed_script;
+
+    CSHA256 hasher;
+    hasher.Write((unsigned char*)&(*script.begin()), script.end() - script.begin());
+    hasher.Finalize(hashed_script.begin());
+
+    return hashed_script.GetUint64(0);
+}
+
+bool AddrIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<uint64_t, CDiskTxPos>> positions;
+    positions.reserve(2 * block.vtx.size());  // Most transactions have at least 1 input and 1 output.
+
+    // Index addresses of spent outputs if txindex is enabled,
+    for (const auto& tx : block.vtx) {
+        for (const auto tx_out : tx->vout){
```

Comment:
```
index/addrindex.cpp:110:25: warning: loop variable 'tx_out' of type 'const CTxOut' creates a copy from type 'const CTxOut' [-Wrange-loop-analysis]
        for (const auto tx_out : tx->vout){
                        ^
index/addrindex.cpp:110:14: note: use reference type 'const CTxOut &' to prevent copying
        for (const auto tx_out : tx->vout){
             ^~~~~~~~~~~~~~~~~~~
```